File: leetcode_python/34.py
from typing import List
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        l = len(nums)
        begin = 0
        while begin < l:
            if nums[begin] == target:
                break
            else:
                begin +=1
        end = begin
        while end < l:
            if nums[end] == target:
                end += 1
            else:
                break 
        if begin == l:
            return [-1,-1]
        return [begin,end-1]
    # 使用二分查找找的一个target 然后再从这个target进行中心扩散
    def searchRange_binarySearch(self,nums:List[int],target:int)-> List[int]:
        l = len(nums)
        left ,right  = 0,l-1
        m = -1
        while left <= right:
            middle = (left+right)//2
            if nums[middle] == target:
                m = middle # 这是中心 
                break 
            if nums[middle] > target:
                right = middle-1
            else:
                left = middle + 1
        # 然后中心扩散
        if m == -1:
            return [-1,-1]
        left , right = m,m
        while True:
            if left >= 0 and nums[left] == target:
                left -= 1
            elif right < l and nums[right] == target:
                right += 1
            else:
                break
        return [left+1,right-1]
                
        
s = Solution()
print(s.searchRange_binarySearch([1,1,1,2,2,3,3,6],1))

File: leetcode_c_plus_plus/34.cpp
// 34. 在排序数组中查找元素的第一个和最后一个位置
// https://leetcode-cn.com/problems/find-first-and-last-position-of-element-in-sorted-array/
#include<map>
#include<set>
#include<vector>
#include<cmath>
#include<numeric>
#include<unordered_map>
using namespace std;
#define INT_MAX 2147483647
#define INT_MIN (-INT_MAX - 1)
/**
@File : 34.cpp
@Time : 2022/01/09 21:59:30
@Author : YuMin Zhang
@State : Indepeedent 
@Thinking :
@Tag : Medium
class Solution {
**/
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        if(nums.size() == 0) return vector<int>{-1, -1};
        int left = 0, right = nums.size();
        if(nums[right - 1] < target) return vector<int>{-1, -1};
        if(target < nums[left]) return vector<int>{-1, -1};
        while(left < right) {
            int mid = left + (right - left) / 2;
            if(nums[mid] >= target) right = mid;
            else left = mid + 1;
        }
        int start = left;
        if(start < 0 || nums[start] != target) start = -1;
        left = 0;
        right = nums.size();
        while(left < right) {
            int mid = left + (right - left) / 2;
            if(nums[mid] > target) right = mid;
            else left = mid + 1;
        }
        int end = left - 1;
        if(end < 0 || nums[end] != target) end = -1;
        return vector<int>{start, end};

    }
};